File: app/src/main/java/io/hudepohl/gitjobs/ui/main/MainPresenter.java
package io.hudepohl.gitjobs.ui.main;

import java.util.List;

import io.hudepohl.github_jobs.data.api.GitHubJobsInteractor;
import io.hudepohl.github_jobs.data.api.model.GitHubJob;

/**
 * Created by trent on 1/11/18.
 */

public class MainPresenter implements GitHubJobsInteractor.Presenter {

    private MainPresenter.View mView;
    private GitHubJobsInteractor mGitHubJobsModel;

    private int page = 1;

    MainPresenter(MainPresenter.View view) {
        mView = view;
        mGitHubJobsModel = new GitHubJobsInteractor(this);
    }

    void init() {
        page = 1;
        mView.showPageLoadingDialog();
        mGitHubJobsModel.getGitHubJobsList(page);
    }

    void refresh() {
        page = 1;
        mGitHubJobsModel.getGitHubJobsList(page);
    }

    void getNextPage() {
        mView.showPageLoadingDialog();
        mGitHubJobsModel.getGitHubJobsList(++page);
    }

    @Override
    public void onGetJobListSuccess(List<GitHubJob> jobList) {
        if (page == 1) {
            mView.initializeJobList(jobList);
        } else {
            mView.addJobsToList(jobList);
        }

        mView.hidePageLoadingDialog();
    }

    @Override
    public void onGetJobListFailure() {
        mView.showGetJobListError();
    }

    @Override
    public void onGetJobSuccess(GitHubJob job) {

    }

    @Override
    public void onGetJobFailure() {

    }

    interface View {

        void initializeJobList(List<GitHubJob> jobs);
        void addJobsToList(List<GitHubJob> jobs);
        void showPageLoadingDialog();
        void hidePageLoadingDialog();
        void showGetJobListError();

    }

}

File: app/src/main/java/io/hudepohl/gitjobs/ui/main/MainPresenter.kt
package io.hudepohl.gitjobs.ui.main

import io.hudepohl.gitjobs.data.githubJobs.GitHubJobsAPI
import io.hudepohl.gitjobs.data.githubJobs.model.GitHubJob
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by trent on 1/11/18.
 */

class MainPresenter @Inject constructor() {

    private var view: MainPresenter.View? = null
    @Inject lateinit var api: GitHubJobsAPI

    fun attachView(viewToAttach: MainPresenter.View) {
        view = viewToAttach
    }

    fun detachView() {
        view = null
    }

    fun init() {

        api.getJobList(1)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<List<GitHubJob>> {

                    override fun onSubscribe(d: Disposable)     { view?.showPageLoadingDialog() }
                    override fun onNext(jobs: List<GitHubJob>)  { view?.initializeJobList(jobs) }
                    override fun onError(e: Throwable)          { view?.showGetJobListError()   }
                    override fun onComplete()                   { view?.hidePageLoadingDialog() }
                })
    }

    fun refresh() {

        api.getJobList(1)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<List<GitHubJob>> {

                    override fun onSubscribe(d: Disposable)     {  } // <-- no dialog on refresh
                    override fun onNext(jobs: List<GitHubJob>)  { view?.initializeJobList(jobs) }
                    override fun onError(e: Throwable)          { view?.showGetJobListError()   }
                    override fun onComplete()                   {  }
                })
    }

    fun nextPage(page: Int) {

        api.getJobList(page)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<List<GitHubJob>> {

                    override fun onSubscribe(d: Disposable)     { view?.showPageLoadingDialog() }
                    override fun onNext(jobs: List<GitHubJob>)  { view?.addJobsToList(jobs)     }
                    override fun onError(e: Throwable)          { view?.showGetJobListError()   }
                    override fun onComplete()                   { view?.hidePageLoadingDialog() }
                })
    }

    interface View {
        fun initializeJobList(jobs: List<GitHubJob>)
        fun addJobsToList(jobs: List<GitHubJob>)
        fun showPageLoadingDialog()
        fun hidePageLoadingDialog()
        fun showGetJobListError()
    }
}
